[torchelastic][c10d] Fix store prefix race in rendezvous

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135957

1. We want to take option 3 as discussed in https://github.com/pytorch/pytorch/issues/135712, so every time when we retry, we create a new TCPStore server first so that we don't need to append attempt count as prefix and avoid eventually TCPStore sync failure. (This is only for the TCPStore sharing enabled case) 
2. We start a new server bound to an ephemeral port (i.e. 0) so it gets assigned to a free port. We then pass that downstream (trainer or c10d). By doing so, TCPStore is managed by the elastic agent rather than having a race condition on binding to a specific port in the trainer.
3. Then the port be broadcasted for dynamic_rendezvous.

Only one more question, what do we do about the store created from (_create_tcp_store) torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py, are we ok with creating a duplicate TCPStore server?

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D63396829](https://our.internmc.facebook.com/intern/diff/D63396829)

fduwjj committed:
[RFC][c10d] Fix store prefix race in rendezvous

[ghstack-poisoned]

fduwjj commented:
Send out the PR for RFC first and unit test is on the way.

fduwjj committed:
Update on "[RFC][torchelastic] Fix store prefix race in rendezvous"


1. We want to take option 3 as discussed in https://github.com/pytorch/pytorch/issues/135712, so every time when we retry, we create a new TCPStore so that we don't need to append attemp prefix and avoid eventually TCPStore sync failure.
2. Upon checking the code, we creating a new TCPStore in c10d_rendezvous_backend.py before we even hitting the the logic inside dynamic_rendezvous.py. Since we use `multi_tenant=True`, so we don't create new server in TCPStore.cpp if this flag is set:
```cpp
if (opts.multiTenant) {
    std::lock_guard<std::mutex> guard{cache_mutex_};

    // If the caller is okay with a multi-tenant store, first check if we
    // already have a TCPServer running on the specified port.
    if (opts.port > 0) {
      auto pos = cachedServers_.find(opts.port);
      if (pos != cachedServers_.end()) {
        server = pos->second.lock();
        if (server != nullptr) {
          return server;
        }

        // Looks like the TCPStore has been disposed, make sure that we release
        // the control block.
        cachedServers_.erase(pos);
      }
    }

    server = startCore();

    cachedServers_.emplace(server->port(), server);
  }
```
so if we are sure the TCPStore server is destroyed everytime we retry, we are recreating a TCPStore server already; otherwise, we need more changes to the elastic code.
3. This changes will force broadcasting a new mast address and port everytime when dynamic_rendezvous's `next_rendezvous` is called.

Send a PR first to collect feedback.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

fduwjj commented on torch/distributed/elastic/rendezvous/api.py:
```diff
@@ -83,7 +83,14 @@ def build(
+                port = store.port
+            else:
+                # Not all stores define the property of port, so we fallback to get_free_port.
+                port = get_free_port()
```
Not sure if we want to create a TCPStore here directly.

d4l3k replied:
I'm not sure this change is safe -- we may want to change the build() interface to pass in an optional port

If we're not using a shared store we shouldn't be passing the shared store port to the child workers. And the shared store path bypasses this function I believe and just directly creates a RendezvousStoreInfo object

d4l3k commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -1188,25 +1188,19 @@ def next_rendezvous(self) -> RendezvousInfo:
+            # bootstrapping info across ranks
+            self._bootstrap_store_info = RendezvousStoreInfo.build(
+                rank, store, local_addr=self._this_node.addr
+            )
```
Can we move this after the rank0 check and instead pass in the port from `_shared_tcp_store_server` on rank0 to be broadcasted?

d4l3k commented on torch/distributed/rendezvous.py:
```diff
@@ -181,9 +181,8 @@ def _create_c10d_store(
-        tcp_store = TCPStore(hostname, port, world_size, False, timeout)
-        return PrefixStore(f"/worker/attempt_{attempt}", tcp_store)
+        # We create a new TCPStore for every retry so no need to add prefix for each attempt.
+        return TCPStore(hostname, port, world_size, False, timeout)
```
pre-existing nit: can we use kwargs for the False argument?

fduwjj committed:
Update on "[RFC][torchelastic][c10d] Fix store prefix race in rendezvous"


1. We want to take option 3 as discussed in https://github.com/pytorch/pytorch/issues/135712, so every time when we retry, we create a new TCPStore so that we don't need to append attemp prefix and avoid eventually TCPStore sync failure.
2. Upon checking the code, we creating a new TCPStore in c10d_rendezvous_backend.py before we even hitting the the logic inside dynamic_rendezvous.py. Since we use `multi_tenant=True`, so we don't create new server in TCPStore.cpp if this flag is set:
```cpp
if (opts.multiTenant) {
    std::lock_guard<std::mutex> guard{cache_mutex_};

    // If the caller is okay with a multi-tenant store, first check if we
    // already have a TCPServer running on the specified port.
    if (opts.port > 0) {
      auto pos = cachedServers_.find(opts.port);
      if (pos != cachedServers_.end()) {
        server = pos->second.lock();
        if (server != nullptr) {
          return server;
        }

        // Looks like the TCPStore has been disposed, make sure that we release
        // the control block.
        cachedServers_.erase(pos);
      }
    }

    server = startCore();

    cachedServers_.emplace(server->port(), server);
  }
```
so if we are sure the TCPStore server is destroyed everytime we retry, we are recreating a TCPStore server already; otherwise, we need more changes to the elastic code.
3. This changes will force broadcasting a new mast address and port everytime when dynamic_rendezvous's `next_rendezvous` is called.

Send a PR first to collect feedback.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

d4l3k commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -1188,25 +1189,31 @@ def next_rendezvous(self) -> RendezvousInfo:
+                # Since we create TCPStore with `multi_tenant = True`, we won't create
+                # a new TCPStore service if there is already one running.
+                if isinstance(self._store, dist.TCPStore):
+                    self._shared_tcp_store_server = self._store
```
I don't think we need this branch anymore

Can just nuke it since we always want to start a clean store

d4l3k commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -1188,25 +1189,31 @@ def next_rendezvous(self) -> RendezvousInfo:
                 self._bootstrap_store_info = RendezvousStoreInfo(
-                    master_addr=addr, master_port=port
+                    master_addr=self._store.host, master_port=self._store.port
                 )
```
we likewise can get rid of this branch

d4l3k reviewed (commented):
looking good -- can we add some test for this?

d4l3k commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -1188,25 +1189,31 @@ def next_rendezvous(self) -> RendezvousInfo:
+                    self._shared_tcp_store_server = self._store
+                else:
+                    self._shared_tcp_store_server = self._create_tcp_store_server(
+                        self._this_node.addr, get_free_port()
```
we should use port=0 instead of `get_free_port()`

fduwjj committed:
Update on "[RFC][torchelastic][c10d] Fix store prefix race in rendezvous"


1. We want to take option 3 as discussed in https://github.com/pytorch/pytorch/issues/135712, so every time when we retry, we create a new TCPStore so that we don't need to append attemp prefix and avoid eventually TCPStore sync failure.
2. Upon checking the code, we creating a new TCPStore in c10d_rendezvous_backend.py before we even hitting the the logic inside dynamic_rendezvous.py. Since we use `multi_tenant=True`, so we don't create new server in TCPStore.cpp if this flag is set:
```cpp
if (opts.multiTenant) {
    std::lock_guard<std::mutex> guard{cache_mutex_};

    // If the caller is okay with a multi-tenant store, first check if we
    // already have a TCPServer running on the specified port.
    if (opts.port > 0) {
      auto pos = cachedServers_.find(opts.port);
      if (pos != cachedServers_.end()) {
        server = pos->second.lock();
        if (server != nullptr) {
          return server;
        }

        // Looks like the TCPStore has been disposed, make sure that we release
        // the control block.
        cachedServers_.erase(pos);
      }
    }

    server = startCore();

    cachedServers_.emplace(server->port(), server);
  }
```
so if we are sure the TCPStore server is destroyed everytime we retry, we are recreating a TCPStore server already; otherwise, we need more changes to the elastic code.
3. This changes will force broadcasting a new mast address and port everytime when dynamic_rendezvous's `next_rendezvous` is called.

Send a PR first to collect feedback.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC][torchelastic][c10d] Fix store prefix race in rendezvous"


1. We want to take option 3 as discussed in https://github.com/pytorch/pytorch/issues/135712, so every time when we retry, we create a new TCPStore so that we don't need to append attemp prefix and avoid eventually TCPStore sync failure.
2. Upon checking the code, we creating a new TCPStore in c10d_rendezvous_backend.py before we even hitting the the logic inside dynamic_rendezvous.py. Since we use `multi_tenant=True`, so we don't create new server in TCPStore.cpp if this flag is set:
```cpp
if (opts.multiTenant) {
    std::lock_guard<std::mutex> guard{cache_mutex_};

    // If the caller is okay with a multi-tenant store, first check if we
    // already have a TCPServer running on the specified port.
    if (opts.port > 0) {
      auto pos = cachedServers_.find(opts.port);
      if (pos != cachedServers_.end()) {
        server = pos->second.lock();
        if (server != nullptr) {
          return server;
        }

        // Looks like the TCPStore has been disposed, make sure that we release
        // the control block.
        cachedServers_.erase(pos);
      }
    }

    server = startCore();

    cachedServers_.emplace(server->port(), server);
  }
```
so if we are sure the TCPStore server is destroyed everytime we retry, we are recreating a TCPStore server already; otherwise, we need more changes to the elastic code.
3. This changes will force broadcasting a new mast address and port everytime when dynamic_rendezvous's `next_rendezvous` is called.

Send a PR first to collect feedback.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

fduwjj commented:
Looks like there are some existing test cases already, we just need to update them.

kwen2501 commented on test/distributed/elastic/rendezvous/dynamic_rendezvous_test.py:
```diff
@@ -1795,14 +1804,27 @@ def test_use_agent_store_is_disabled(self):
-        prefix_store = Mock(spec=dist.PrefixStore)
-        prefix_store_class_mock.return_value = prefix_store
+        expected_addr = "expected_address"
+        expected_port = 54231
```
Not related to this PR, maybe define some constants so that later comers do not need to search for each instance of 54321 if they want to modify it?

fduwjj replied:
I agree that we will need to make it constant and I will do that. But this change does not fall into that you said, even with the constant, we are not changing it everywhere. This is an expected behavior change.

fduwjj committed:
Update on "[RFC][torchelastic][c10d] Fix store prefix race in rendezvous"


1. We want to take option 3 as discussed in https://github.com/pytorch/pytorch/issues/135712, so every time when we retry, we create a new TCPStore server first so that we don't need to append attempt count as prefix and avoid eventually TCPStore sync failure. (This is only for the TCPStore sharing enabled case) 
2. We reuse the port from the newly created TCPStore server and first bind it to port 0 so that the system will allocate a new available one to us.
3. Then the port be broadcasted for dynamic_rendezvous.

Only one more question, what do we do about the store created from (_create_tcp_store) torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py, are we ok with creating a duplicate TCPStore server?

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

kwen2501 commented:
> We reuse the port from the newly created TCPStore server and first bind it to port 0 so that the system will allocate a new available one to us.
I don't quite understand the point between reuse and allocate new one.

d4l3k reviewed (approved):
LGTM! Thanks for updating the tests :)

c-p-i-o commented on test/distributed/elastic/rendezvous/dynamic_rendezvous_test.py:
```diff
@@ -1217,19 +1223,30 @@ def test_share_store_creates_tcp_store(self):
-            type(handler._store).host = PropertyMock(return_value="host")
-            type(handler._store).port = PropertyMock(return_value=54321)
+            type(handler._store).host = PropertyMock(return_value=TEST_ADDR)
+            type(handler._store).port = PropertyMock(return_value=54320)
```
Is this `TEST_PORT - 1` always?

fduwjj replied:
yes. let me change that.

c-p-i-o commented on test/distributed/elastic/rendezvous/dynamic_rendezvous_test.py:
```diff
@@ -1795,14 +1812,27 @@ def test_use_agent_store_is_disabled(self):
-        prefix_store = Mock(spec=dist.PrefixStore)
-        prefix_store_class_mock.return_value = prefix_store
+        expected_addr = "expected_address"
+        expected_port = 54231
```
Shouldn't this be `54321` a.k.a `TEST_PORT` and not `54231`?

fduwjj replied:
no, this is indeed expected because we are create a new TCPStore anyway.

d4l3k commented:
@kwen2501 I think that comment is slightly out of date -- in this PR we start a new server bound to an ephemeral port (i.e. 0) so it gets assigned to a free port. We then pass that downstream. We don't reuse the server or ports here. It's not really "shared" it's more that the TCPStore is managed by the elastic agent rather than having a race condition on binding to a specific port in the trainer

c-p-i-o reviewed (approved):
Minor comments in the test section - but LGTM overall!

fduwjj commented:
@kwen2501 yes, the port is for passing downstream to trainer or c10d. I updated the PR descriptions to make it more accurate.

fduwjj committed:
Update on "[RFC][torchelastic][c10d] Fix store prefix race in rendezvous"


1. We want to take option 3 as discussed in https://github.com/pytorch/pytorch/issues/135712, so every time when we retry, we create a new TCPStore server first so that we don't need to append attempt count as prefix and avoid eventually TCPStore sync failure. (This is only for the TCPStore sharing enabled case) 
2. We start a new server bound to an ephemeral port (i.e. 0) so it gets assigned to a free port. We then pass that downstream (trainer or c10d). By doing so, TCPStore is managed by the elastic agent rather than having a race condition on binding to a specific port in the trainer.
3. Then the port be broadcasted for dynamic_rendezvous.

Only one more question, what do we do about the store created from (_create_tcp_store) torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py, are we ok with creating a duplicate TCPStore server?

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj committed:
Update on "[torchelastic][c10d] Fix store prefix race in rendezvous"


1. We want to take option 3 as discussed in https://github.com/pytorch/pytorch/issues/135712, so every time when we retry, we create a new TCPStore server first so that we don't need to append attempt count as prefix and avoid eventually TCPStore sync failure. (This is only for the TCPStore sharing enabled case) 
2. We start a new server bound to an ephemeral port (i.e. 0) so it gets assigned to a free port. We then pass that downstream (trainer or c10d). By doing so, TCPStore is managed by the elastic agent rather than having a race condition on binding to a specific port in the trainer.
3. Then the port be broadcasted for dynamic_rendezvous.

Only one more question, what do we do about the store created from (_create_tcp_store) torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py, are we ok with creating a duplicate TCPStore server?

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

fduwjj committed:
Update on "[torchelastic][c10d] Fix store prefix race in rendezvous"


1. We want to take option 3 as discussed in https://github.com/pytorch/pytorch/issues/135712, so every time when we retry, we create a new TCPStore server first so that we don't need to append attempt count as prefix and avoid eventually TCPStore sync failure. (This is only for the TCPStore sharing enabled case) 
2. We start a new server bound to an ephemeral port (i.e. 0) so it gets assigned to a free port. We then pass that downstream (trainer or c10d). By doing so, TCPStore is managed by the elastic agent rather than having a race condition on binding to a specific port in the trainer.
3. Then the port be broadcasted for dynamic_rendezvous.

Only one more question, what do we do about the store created from (_create_tcp_store) torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py, are we ok with creating a duplicate TCPStore server?

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63396829).

izaitsevfb commented:
@pytorchbot merge -f 'landed internally'

atalman commented:
@pytorchbot merge -f 'landed internally. one more try'

atalman commented:
Please note. This change was reverted internally. I suggest we close this PR and open a new one if we want to land this change

fduwjj commented:
recreate a PR in https://github.com/pytorch/pytorch/pull/136768
